pkg/apis/kubeone/helpers.go

```diff
@@ -294,77 +294,92 @@ func (p CloudProviderSpec) CloudProviderInTree() bool {
-func (c KubeOneCluster) InTreePluginUnregisterFeatureGate() []string {
-	lessThan21, _ := semver.NewConstraint("< 1.21.0")
-	ver, _ := semver.NewVersion(c.Versions.Kubernetes)
+	lessThan21 := lessThan21Constraint.Check(ver)
+	featureGates := map[string]bool{}
 
 	switch {
+	case c.CloudProvider.AWS != nil:
+		featureGates["CSIMigrationAWS"] = true
+		if complete {
+			if lessThan21 {
+				featureGates["CSIMigrationAWSComplete"] = true
+			} else {
+				featureGates["InTreePluginAWSUnregister"] = true
+			}
+		}
 	case c.CloudProvider.Azure != nil:
-		if lessThan21.Check(ver) {
-			return []string{"CSIMigrationAzureDiskComplete", "CSIMigrationAzureFileComplete"}
+		featureGates["CSIMigrationAzureDisk"] = true
+		featureGates["CSIMigrationAzureFile"] = true
+		if complete {
+			if lessThan21 {
+				featureGates["CSIMigrationAzureDiskComplete"] = true
+				featureGates["CSIMigrationAzureFileComplete"] = true
+			} else {
+				featureGates["InTreePluginAzureDiskUnregister"] = true
+				featureGates["InTreePluginAzureFileUnregister"] = true
+			}
 		}
-
-		return []string{"InTreePluginAzureDiskUnregister", "InTreePluginAzureFileUnregister"}
 	case c.CloudProvider.Openstack != nil:
-		if lessThan21.Check(ver) {
-			return []string{"CSIMigrationOpenStackComplete"}
+		featureGates["CSIMigrationOpenStack"] = true
+		featureGates["ExpandCSIVolumes"] = true
+		if complete {
+			if lessThan21 {
+				featureGates["CSIMigrationOpenStackComplete"] = true
+			} else {
+				featureGates["InTreePluginOpenStackUnregister"] = true
+			}
 		}
-
-		return []string{"InTreePluginOpenStackUnregister"}
 	case c.CloudProvider.Vsphere != nil:
-		if lessThan21.Check(ver) {
-			return []string{"CSIMigrationvSphereComplete"}
+		featureGates["CSIMigrationvSphere"] = true
+		if complete {
+			if lessThan21 {
+				featureGates["CSIMigrationvSphereComplete"] = true
+			} else {
+				featureGates["InTreePluginvSphereUnregister"] = true
+			}
 		}
+	default:
+		return nil, fail.ConfigValidation(fmt.Errorf("csi migration is not supported for selected provider"))
+	}
 
-		return []string{"InTreePluginvSphereUnregister"}
+	return featureGates, nil
+}
+
+// CSIMigrationFeatureGates returns CSI migration feature gates in form of a map
+// (to be used with Kubelet config) and string (to be used with kube-apiserver
+// and kube-controller-manager)
+// NB: We're intentionally not enabling CSIMigration feature gate because it's
+// enabled by default since Kubernetes 1.18
+// (https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/)
+// This is a KubeOneCluster function because feature gates are Kubernetes-version dependent.
+func (c KubeOneCluster) CSIMigrationFeatureGates(complete bool) (map[string]bool, string, error) {
+	featureGates, err := c.csiMigrationFeatureGates(complete)
+	if err != nil {
+		return nil, "", err
 	}
 
-	return nil
+	return featureGates, marshalFeatureGates(featureGates), nil
 }
 
 func marshalFeatureGates(fgm map[string]bool) string {
```

Comment:
changes from above in whole file is done in the name of having a single place where CSI featureGates are defined